synthtool/metadata.py

```diff
@@ -224,38 +224,110 @@ def get_touched_file_paths(self) -> List[str]:
 class MetadataTrackerAndWriter:
     """Writes metadata file upon exiting scope."""
 
+    # _thread_local.live_tracker points to the MetadataTrackerAndWriter in the current
+    #  thread context.
+    _thread_local = threading.local()
+
     def __init__(self, metadata_file_path: str):
         self.metadata_file_path = metadata_file_path
+        self.generated_file_paths: List[str] = []
 
     def __enter__(self):
+        self.old_tracker = MetadataTrackerAndWriter.get_live_tracker()
+        self._thread_local.live_tracker = self
         self.old_metadata = _read_or_empty(self.metadata_file_path)
         _add_self_git_source()
-        watch_dir = pathlib.Path(self.metadata_file_path).parent
-        os.makedirs(watch_dir, exist_ok=True)
-        self.handler = FileSystemEventHandler(watch_dir)
-        self.observer = watchdog.observers.Observer()
-        self.observer.schedule(self.handler, str(watch_dir), recursive=True)
-        self.observer.start()
+        self.start_tracking()
+
+    @staticmethod
+    def get_live_tracker():
+        """Gets the MetadataTrackerAndWriter in the current thread context.
+
+        Returns None if there is none in the current thread context."""
+        return getattr(MetadataTrackerAndWriter._thread_local, "live_tracker", None)
+
+    def clear_tracking_history(self):
+        """Clears the tracking history of generated files."""
+        self.generated_file_paths = []
+
+    def start_tracking(self):
+        """Create a watchdog observer and start tracking file changes."""
+        observer = getattr(self, "observer", None)
+        if observer:
+            pass  # Already tracking.  Keep tracking.
+        else:
+            watch_dir = pathlib.Path(self.metadata_file_path).parent
+            os.makedirs(watch_dir, exist_ok=True)
+            self.handler = FileSystemEventHandler(watch_dir)
+            self.observer = watchdog.observers.Observer()
+            self.observer.schedule(self.handler, str(watch_dir), recursive=True)
+            self.observer.start()
+
+    def stop_tracking(self):
+        """Shut down existing watchdog observer.  Does nothing if not observing."""
+        observer = getattr(self, "observer", None)
+        if observer:
+            time.sleep(2)  # Finish collecting observations about modified files.
+            observer.stop()
+            observer.join()
```

Comment:
If I'm following this correctly, the handler runs in its own thread, and `join` will basically bring it in to our current thread allowing us access to the list it has populated?